### PR TITLE
fix: do not split quoted key as path

### LINKED ISSUE #269: Substitution-like strings in keys are parsed incorrectly
According to the [HOCON spec](https://github.com/lightbend/config/blob/5d6a46631c7ca1d617de02459bd62af5b875073b/HOCON.md#substitutions):

> Substitutions are not parsed inside quoted strings.

### Expected behavior

```elixir
iex> :hocon.binary(~s|"${.x}" = 1|)      
{:ok, %{"${.x}" => 1}}
```

### Actual behavior

```elixir
iex> :hocon.binary(~s|"${.x}" = 1|)      
{:ok, %{"${" => %{"x}" => 1}}}
```

### PR DESCRIPTION
**Pull request by zmstone**
fixes #269

### CODE REVIEW

File: src/hocon_token.erl [hunk 1]
```diff
@@ -219,7 +222,12 @@ abspath(Var, PathStack) ->
 do_abspath(Var, ['$root']) ->
     Var;
 do_abspath(Var, [#{?HOCON_T := key} = K | More]) ->
-    do_abspath(unicode_bin([value_of(K), <<".">>, Var]), More).
+    do_abspath(unicode_bin([maybe_join(value_of(K)), <<".">>, Var]), More).
+
+maybe_join({keypath, Path}) ->
+    infix(Path, ".");
```
Comment (thalesmg) [inline src/hocon_token.erl:228]:
```suggestion
    lists:join(".", Path);
```

File: src/hocon_token.erl [hunk 2]
```diff
@@ -302,3 +310,12 @@ format_error(Line, ErrorInfo, Ctx) ->
...
 unicode_list(B) -> unicode:characters_to_list(B, utf8).
+
+paths(Key) when is_binary(Key) ->
+    paths(unicode:characters_to_list(Key, utf8));
+paths(Key) when is_list(Key) ->
+    lists:map(fun unicode_bin/1, string:tokens(Key, ".")).
+
+infix([], _) -> [];
+infix([X], _) -> [X];
+infix([H | T], I) -> [H, I | infix(T, I)].
```
Comment (thalesmg) [inline src/hocon_token.erl:321]:
```suggestion
```

**Review by thalesmg — APPROVED** [review]
I suggest adding the following test case:

```erlang
map_with_placeholders_test() ->
    RawConf =
        #{ <<"headers">> =>
               #{
                 <<"fixed_key">> => <<"fixed_value">>,
                 <<"${.payload.key}">> => <<"fixed_value">>,
                 <<"${.payload.key}2">> => <<"${.payload.value}">>,
                 <<"fixed_key2">> => <<"${.payload.value}">>
                }
         },
    TmpFile = "/tmp/" ++ atom_to_list(?FUNCTION_NAME) ++ ".conf",
    try
        ok = file:write_file(TmpFile, hocon_pp:do(RawConf, #{})),
        {ok, LoadedConf} = hocon:load(TmpFile, #{format => map}),
        ?assertEqual(RawConf, LoadedConf),
        ok
    after
        file:delete(TmpFile)
    end.
```

**Review by thalesmg — APPROVED** [review]

File: src/hocon_pp.erl [hunk 3]
```diff
@@ -184,7 +184,7 @@ is_quote_key(K) ->
 
 is_quote_str(S) ->
     case hocon_scanner:string(S) of
-        {ok, [{string, 1, S}], 1} ->
+        {ok, [{Tag, 1, S}], 1} when Tag =:= string orelse Tag =:= unqstr ->
```
Comment (thalesmg) [inline src/hocon_pp.erl:187]: Q: the function name, at first glance, seems to indicate whether we have a quoted string, but here we are matching unquoted strings?  Perhaps the fn purpose is to decide whether to quote the input?

**Review by thalesmg — APPROVED** [review]

**Comment by zmstone** [comment]
tagged: `0.39.17`

**Pull request merged**